src/Illuminate/View/Concerns/CachesComponents.php

```diff
@@ -0,0 +1,126 @@
+<?php
+
+namespace Illuminate\View\Concerns;
+
+use Illuminate\Container\Container;
+use Illuminate\Contracts\Support\Htmlable;
+use Illuminate\Contracts\View\CanBeCached;
+use Illuminate\Contracts\View\View;
+use Illuminate\Support\Facades\Cache;
+
+trait CachesComponents
+{
+    /**
+     * The cache controlling attributes for this component.
+     *
+     * @return array
+     */
+    protected function cacheAttributes()
+    {
+        return [
+            'x-cache',
+            'x-cache-key',
+            'x-cache-ttl',
+        ];
+    }
+
+    /**
+     * Determines if the component can be cached.
+     *
+     * @return bool
+     */
+    protected function canBeCached()
+    {
+        return $this instanceof CanBeCached;
+    }
+
+    /**
+     * Determines if the component should be cached.
+     *
+     * @return bool
+     */
+    protected function shouldBeCached()
+    {
+        return $this->attributes->get('x-cache', false) && $this->canBeCached();
+    }
+
+    /**
+     * The key used to cache the component.
+     *
+     * @param  array  $slots
+     * @return bool
+     */
+    protected function cacheKey(array $slots)
+    {
+        return $this->attributes->get('x-cache-key', static::class.sha1($this->attributes->except($this->cacheAttributes())->toHtml().serialize($slots)));
+    }
+
+    /**
+     * The cache TTL to use for the component.
+     *
+     * @return mixed
+     */
+    protected function cacheTtl()
+    {
+        return $this->attributes->get('x-cache-ttl', now()->addHour());
+    }
+
+    /**
+     * Render the resolved component view so that it can be cached.
+     *
+     * @param  \Illuminate\Contracts\View\View|\Illuminate\Contracts\Support\Htmlable|string $view
+     * @param  array  $data
+     * @return string
+     *
+     * @throws \Illuminate\Contracts\Container\BindingResolutionException
+     */
+    protected function renderResolvedView($view, $data)
```

Comment:
Shouldn't this handle when a class based component returns a `\Closure`?

On the last `if` branch from the `Component@resolveView` a `Closure` `$resolvedView` can be a closure.

https://github.com/laravel/framework/pull/36928/files#diff-1005052c893d685dfa5b9c5c7431e2a880260f9a16cba75e16689ec33549755cR85-R88

I followed the track into here and didn't figure out how a `Closure` is handled.

I guess this would suffice:

~~~php
if ($view instance of Closure) {
    return $view($data);
}
~~~

Apologies if I missed something silly.